src/modules/resource/drain.c

```diff
@@ -0,0 +1,302 @@
+            (void)snprintf (errbuf, sizeof (errbuf), "rank %u not drained", id);
+            errno = EINVAL;
+            errstr = errbuf;
+            goto error;
+        }
+        id = idset_next (idset, id);
+    }
+    if (rutil_idset_sub (drain->idset, idset) < 0)
+        goto error;
+    if (reslog_post_pack (drain->ctx->reslog,
+                          msg,
+                          "undrain",
+                          "{s:s}",
+                          "idset",
+                          s) < 0) {
+        int saved_errno = errno;
+        (void)rutil_idset_add (drain->idset, idset); // restore orig.
+        errno = saved_errno;
+        goto error;
+    }
+    idset_destroy (idset);
+    return;
+error:
+    if (flux_respond_error (h, msg, errno, errstr) < 0)
+        flux_log_error (h, "error responding to undrain request");
+    idset_destroy (idset);
+}
+
+/* Recover drained idset from eventlog.
+ */
+static int replay_eventlog (struct drain *drain, const json_t *eventlog)
+{
+    size_t index;
+    json_t *entry;
+
+    if (eventlog) {
+        json_array_foreach (eventlog, index, entry) {
+            const char *name;
+            json_t *context;
+            const char *s;
+            struct idset *idset;
+
+            if (eventlog_entry_parse (entry, NULL, &name, &context) < 0)
+                return -1;
+            if (!strcmp (name, "resource-init")) {
+                if (json_unpack (context, "{s:s}", "drain", &s) < 0) {
+                    errno = EPROTO;
+                    return -1;
+                }
+                idset_destroy (drain->idset);
+                if (!(drain->idset = idset_decode (s)))
+                    return -1;
+            }
+            else if (!strcmp (name, "drain")) {
+                if (json_unpack (context, "{s:s}", "idset", &s) < 0) {
+                    errno = EPROTO;
+                    return -1;
+                }
+                if (!(idset = idset_decode (s)))
+                    return -1;
+                if (rutil_idset_add (drain->idset, idset) < 0) {
+                    idset_destroy (idset);
+                    return -1;
+                }
+                idset_destroy (idset);
+            }
+            else if (!strcmp (name, "undrain")) {
+                if (json_unpack (context, "{s:s}", "idset", &s) < 0) {
+                    errno = EPROTO;
+                    return -1;
+                }
+                if (!(idset = idset_decode (s)))
+                    return -1;
+                if (rutil_idset_sub (drain->idset, idset) < 0) {
+                    idset_destroy (idset);
+                    return -1;
+                }
+                idset_destroy (idset);
+            }
+        }
+    }
+    return 0;
+}
+
+static const struct flux_msg_handler_spec htab[] = {
+    { FLUX_MSGTYPE_REQUEST,  MODULE_NAME ".drain", drain_cb, 0 },
+    { FLUX_MSGTYPE_REQUEST,  MODULE_NAME ".undrain", undrain_cb, 0 },
```

Comment:
That would have been fine.  If many messages were going to be parsed by the module, there might be a small advantage to avoiding the glob, since `dispatch_message()` (internal to msg_handler.c) uses a hash to look up request handlers by topic string, then if that fails, it does a linear search through a list where globs are matched with `flux_msg_cmp()`.